maltoolbox/attackgraph/attackgraph.py

```diff
@@ -493,6 +333,176 @@ def attach_attackers(self) -> None:
+                lh_targets = self._follow_expr_chain(
+                    model,
+                    target_assets,
+                    expr_chain.left_link
+                )
+                rh_targets = self._follow_expr_chain(
+                    model,
+                    target_assets,
+                    expr_chain.right_link
+                )
+
+                match (expr_chain.type):
+                    # Once the assets become hashable set operations should be
+                    # used instead.
+                    case 'union':
+                        new_target_assets = lh_targets
+                        for ag_node in rh_targets:
+                            if ag_node not in new_target_assets:
+                                new_target_assets.append(ag_node)
+
+                    case 'intersection':
+                        new_target_assets = []
+                        for ag_node in lh_targets:
+                            if ag_node in rh_targets:
+                                new_target_assets.append(ag_node)
+
+                    case 'difference':
+                        new_target_assets = lh_targets
+                        for ag_node in rh_targets:
+                            if ag_node in new_target_assets:
+                                new_target_assets.remove(ag_node)
+
+                return new_target_assets
+
+            case 'field':
+                # Change the target assets from the current ones to the
+                # associated assets given the specified field name.
+                if not expr_chain.fieldname:
+                    raise LanguageGraphException('"field" step expression '
+                        'chain is missing fieldname.')
+                new_target_assets = []
+                for target_asset in target_assets:
+                    new_target_assets.extend(model.\
+                        get_associated_assets_by_field_name(target_asset,
+                            expr_chain.fieldname))
+                return new_target_assets
+
+            case 'transitive':
+                if not expr_chain.sub_link:
+                    raise LanguageGraphException('"transitive" step '
+                        'expression chain is missing sub link.')
+                accumulated_target_assets = list(target_assets)
+                new_target_assets = list(target_assets)
+                while new_target_assets:
+                    additional_assets = self._follow_expr_chain(
+                        model, new_target_assets, expr_chain.sub_link)
+
+                    new_target_assets = []
+                    for additional_asset in additional_assets:
+                        if additional_asset not in accumulated_target_assets:
+                            accumulated_target_assets.append(additional_asset)
+                            new_target_assets.append(additional_asset)
+
+                return accumulated_target_assets
```

Comment:
I found this block a bit confusing.

```suggestion
                accumulated_target_assets = [target_assets]
                for asset_batch in accumulated_target_assets:
                    additional_assets = self._follow_expr_chain(
                        model, asset_batch, expr_chain.sub_link)

                    # The need for new_assets will be gone once things are sets
                    new_assets = []
                    for additional_asset in additional_assets:
                        if additional_asset not in asset_batch:
                            new_assets.append(additional_asset)
                     
                     if new_assets:
                         accumulated_target_assets.append(new_assets)

                return [asset for batch in accumulated_target_assets for asset in batch]
```

Reply:
Isn't this the transitive block? It needs to keep running until it no longer finds new assets and I don't like having for loops that vary in size over iteration.

Reply:
It was just a suggestion with what seemed to me easier to follow code. Ignore this if you think it adds nothing.